clientv3/grpc_dial_test.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2017 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clientv3
+
+import (
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/coreos/etcd/pkg/testutil"
+
+	"google.golang.org/grpc"
+)
+
+// TestBalancerDoNotBlockOnClose ensures that balancer and grpc don't deadlock each other
+// due to rapid open/close conn. The deadlock causes balancer.Close() to block forever.
+// See issue: https://github.com/coreos/etcd/issues/7283 for more detail.
+func TestBalancerDoNotBlockOnClose(t *testing.T) {
+	defer testutil.AfterTest(t)
+
+	kcl := NewKillConnListener(2, t)
+	defer kcl.Close()
+
+	for i := 0; i < 100; i++ {
+		sb := newSimpleBalancer(kcl.GetEndPoints())
+		conn, err := grpc.Dial("", grpc.WithInsecure(),
+			grpc.WithBalancer(sb))
+		if err != nil {
+			t.Fatal(err)
+		}
+		<-sb.ConnectNotify()
+		closec := make(chan struct{})
+		// Close() might block on balancer.Close()
+		// if balancer and grpc deadlock each other.
+		go func() {
+			defer close(closec)
+			conn.Close()
+		}()
+		select {
+		case <-closec:
+		case <-time.After(time.Second):
+			t.Fatal("conn.Close() timeout")
+		}
+	}
+}
+
+// killConnListener listens incoming conn and kills it immediately.
+type killConnListener struct {
+	wg        sync.WaitGroup
+	eps       []string
+	stopc     chan struct{}
+	bindAddrc chan string
+	t         *testing.T
+}
+
+func NewKillConnListener(size int, t *testing.T) *killConnListener {
+	kcl := &killConnListener{stopc: make(chan struct{}),
+		bindAddrc: make(chan string, size),
+		t:         t}
+	kcl.start(size)
+	return kcl
+}
+
+func (kcl *killConnListener) start(size int) {
+	for i := 0; i < size; i++ {
+		kcl.wg.Add(1)
+		go func() {
+			defer kcl.wg.Done()
+			kcl.listen()
+		}()
+		addr := <-kcl.bindAddrc
+		kcl.eps = append(kcl.eps, addr)
+	}
+}
+
+func (kcl *killConnListener) GetEndPoints() []string {
+	return kcl.eps
+}
+
+func (kcl *killConnListener) listen() {
+	ln, err := net.Listen("tcp", ":0")
+	if err != nil {
+		kcl.t.Fatal(err)
+	}
+	kcl.bindAddrc <- ln.Addr().String()
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
```

Comment:
`listen` is already in a goroutine, there's no gain in concurrency by having another one